src/tpm2_wrap.c

```diff
@@ -7029,837 +7029,323 @@ int wolfTPM2_GetPolicyDigest(TPM_HANDLE sessionHandle, byte* policyDigest,
-
-            if (sigAlg == TPM_ALG_NULL)
-                sigAlg = TPM_ALG_RSASSA;
-        }
-
-        rc = wolfTPM2_SignHashScheme_ex(dev, (WOLFTPM2_KEY*)authKey, aHash,
-            aHashSz, policySignedSig, (int*)policySignedSigSz, sigAlg,
-            policyHashAlg, &policySignedIn->auth);
-    }
-
-    /* add PolicySigned to the policy */
-    if (rc == 0) {
-        policySignedIn->authObject = authKey->handle.hndl;
-
-        /* set the session */
-        policySignedIn->policySession = session->handle.hndl;
-
-        /* set nonce if it's non null */
-        if (nonceSz > 0) {
-            policySignedIn->policyRef.size = nonceSz;
-            XMEMCPY(policySignedIn->policyRef.buffer, nonce, nonceSz);
-        }
-
-        rc = TPM2_PolicySigned(policySignedIn, policySignedOut);
-    }
-
-    /* get the policyDigest */
-    if (rc == 0) {
-        rc = wolfTPM2_GetPolicyDigest(session->handle.hndl, policyDigest,
-            &policyDigestSz);
-    }
-
-    /* get the nvAttributes */
-    if (rc == 0) {
-        parent->hndl = TPM_RH_OWNER;
-
-        rc = wolfTPM2_GetNvAttributesTemplate(parent->hndl, &nvAttributes);
-    }
-
-    /* create the policy digest nv space */
-    if (rc == 0) {
-        rc = wolfTPM2_NVCreateAuth(dev, parent, nv, policyDigestNvIndex,
-            nvAttributes, sizeof(TPM_ALG_ID) + policyDigestSz, NULL, 0);
-
-        if (rc == TPM_RC_NV_DEFINED)
-            rc = 0;
-    }
+    inSz = *digestSz; /* capture input digest size (for approvedPolicy) */
+    rc = TPM2_GetHashType(pcrAlg);
+    hashType = (enum wc_HashType)rc;
+    rc = wc_HashGetDigestSize(hashType);
+    if (rc < 0)
+        return rc;
+    *digestSz = rc; /* set actual size */
 
-    /* write the hash alg as the first two octects, big endian */
-    if (rc == 0) {
-        bigEnd = XHTONS(policyHashAlg);
+    rc = wc_HashInit(&hash_ctx, hashType);
+    if (rc != 0)
+        return rc;
 
-        rc = wolfTPM2_NVWriteAuth(dev, nv, policyDigestNvIndex, (byte*)&bigEnd,
-            sizeof(TPM_ALG_ID), 0);
+    /* approvedPolicy */
+    if (rc == 0 && inSz > 0) {
+        rc = wc_HashUpdate(&hash_ctx, hashType, digest, inSz);
     }
-
-    /* write the policy digest to the nv space */
-    if (rc == 0) {
-        rc = wolfTPM2_NVWriteAuth(dev, nv, policyDigestNvIndex, policyDigest,
-            policyDigestSz, sizeof(TPM_ALG_ID));
+    /* policyRef */
+    if (rc == 0 && policyRefSz > 0) {
+        rc = wc_HashUpdate(&hash_ctx, hashType, digest, inSz);
     }
-
-    /* authorize the policy */
     if (rc == 0) {
-        /* set session */
-        policyAuthNVIn->policySession = session->handle.hndl;
-
-        /* set nvIndex */
-        policyAuthNVIn->nvIndex = policyDigestNvIndex;
-
-        /* set authHandle */
-        policyAuthNVIn->authHandle = parent->hndl;
-
-        rc = TPM2_PolicyAuthorizeNV(policyAuthNVIn);
+        rc = wc_HashFinal(&hash_ctx, hashType, digest);
```

Comment:
this will overwrite the contents of digest, might not be wise if the caller needs the buffer's contents

Reply:
I designed it this way. The "digest" are in both input and output (as documented). This is because some policies take in an "old" digest to get a new one (chaining them).